tests/common/snappi_tests/traffic_generation.py

```diff
@@ -518,25 +520,99 @@ def verify_in_flight_buffer_pkts(duthost,
 
 
-def verify_pause_frame_count(duthost,
-                             snappi_extra_params):
+def verify_pause_frame_count_dut(duthost,
+                                 test_traffic_pause,
+                                 snappi_extra_params):
     """
-    Verify correct frame count for pause frames when the traffic is expected to be paused
+    Verify correct frame count for pause frames when the traffic is expected to be paused or not
+    on the DUT
 
     Args:
         duthost (obj): DUT host object
+        test_traffic_pause (bool): whether test traffic is expected to be paused
         snappi_extra_params (SnappiTestParams obj): additional parameters for Snappi traffic
     Returns:
 
     """
     dut_port_config = snappi_extra_params.base_flow_config["dut_port_config"]
     pytest_assert(dut_port_config is not None, 'Flow port config is not provided')
 
-    for peer_port, prios in dut_port_config[1].items():
+    for peer_port, prios in dut_port_config[0].items():  # TX PFC pause frames
+        for prio in range(len(prios)):
+            pfc_pause_tx_frames = get_pfc_frame_count(duthost, peer_port, prios[prio], is_tx=True)
+            if test_traffic_pause:
+                pytest_assert(pfc_pause_tx_frames > 0,
+                              "PFC pause frames should be transmitted and counted in TX PFC counters for priority {}"
+                              .format(prios[prio]))
+            else:
+                pytest_assert(pfc_pause_tx_frames <= PFC_FRAME_COUNT_TOL,
+                              "PFC pause frames should not be transmitted and counted in TX PFC counters")
+
+    for peer_port, prios in dut_port_config[1].items():  # RX PFC pause frames
         for prio in range(len(prios)):
-            pfc_pause_rx_frames = get_pfc_frame_count(duthost, peer_port, prios[prio])
-            pytest_assert(pfc_pause_rx_frames > 0,
-                          "PFC pause frames with zero source MAC are not counted in the PFC counters")
+            pfc_pause_rx_frames = get_pfc_frame_count(duthost, peer_port, prios[prio], is_tx=False)
+            if test_traffic_pause:
+                pytest_assert(pfc_pause_rx_frames > 0,
+                              "PFC pause frames should be received and counted in RX PFC counters for priority {}"
+                              .format(prios[prio]))
+            else:
+                pytest_assert(pfc_pause_rx_frames <= PFC_FRAME_COUNT_TOL,
+                              "PFC pause frames should not be received and counted in RX PFC counters")
+
+
+def verify_tx_frame_count_dut(duthost,
+                              snappi_extra_params,
+                              tx_frame_count_deviation=0.01,
+                              tx_drop_frame_count_tol=5):
+    """
+    Verify correct frame count for tx frames on the DUT
+    (OK and DROPS) when the traffic is expected to be paused on the DUT.
+    DUT is polled after it stops receiving PFC pause frames from TGEN.
```

Comment:
When you poll DUT, has DUT drained all the packets in the buffer?

Reply:
Yes